src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSFunctionBinding.cs

```diff
@@ -286,18 +405,39 @@ internal static unsafe JSFunctionBinding BindManagedFunctionImpl(string fullyQua
             return signature;
         }
 
+#if !FEATURE_WASM_THREADS
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static unsafe void ResolveOrRejectPromise(Span<JSMarshalerArgument> arguments)
         {
             fixed (JSMarshalerArgument* ptr = arguments)
             {
-                Interop.Runtime.ResolveOrRejectPromise(ptr);
+                Interop.Runtime.ResolveOrRejectPromise((nint)ptr);
                 ref JSMarshalerArgument exceptionArg = ref arguments[0];
                 if (exceptionArg.slot.Type != MarshalerType.None)
                 {
                     JSHostImplementation.ThrowException(ref exceptionArg);
                 }
             }
         }
+#else
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe void ResolveOrRejectPromise(JSProxyContext targetContext, Span<JSMarshalerArgument> arguments)
+        {
+            // this copy is freed in mono_wasm_invoke_import_async
```

Comment:
this used to be always synchronous, and now becomes always-asynchronous. should we preserve the old synchronous execution when the target context is the current context?

Reply:
I think that JS `Promise` handlers always go through microtask queue. So in effect this was already async before from user code perspective. We could avoid allocating C# queue item if we do the optimization you suggest.

I want to switch this to emscripten dispatch later anyway. I will add comment and keep this open question for now.